Flaky test_external_http_authenticator/test.py: Echo server is not responding

https://play.clickhouse.com/play?user=play&run=1#V0lUSAogICAgOTAgQVMgaW50ZXJ2YWxfZGF5cwpTRUxFQ1QKICAgIHRvU3RhcnRPZkRheShjaGVja19zdGFydF90aW1lKSBBUyBkYXksCiAgICBjb3VudCgpIEFTIGZhaWx1cmVzLAogICAgZ3JvdXBVbmlxQXJyYXkocHVsbF9yZXF1ZXN0X251bWJlcikgQVMgcHJzLAogICAgYW55KHJlcG9ydF91cmwpIEFTIHJlcG9ydF91cmwKRlJPTSBjaGVja3MKV0hFUkUgKG5vdygpIC0gdG9JbnRlcnZhbERheShpbnRlcnZhbF9kYXlzKSkgPD0gY2hlY2tfc3RhcnRfdGltZQogICAgQU5EIHRlc3RfbmFtZSA9ICd0ZXN0X2V4dGVybmFsX2h0dHBfYXV0aGVudGljYXRvci90ZXN0LnB5Ojp0ZXN0X3VzZXJfZnJvbV9jb25maWdfYmFzaWNfYXV0aF9wYXNzJwogICAgLS0gQU5EIGNoZWNrX25hbWUgPSAnSW50ZWdyYXRpb24gdGVzdHMgKGFtZF9hc2FuLCBvbGQgYW5hbHl6ZXIsIDIvNiknCiAgICBBTkQgdGVzdF9zdGF0dXMgSU4gKCdGQUlMJywgJ0VSUk9SJykKICAgIEFORCAoKGhlYWRfcmVmID0gJ21hc3RlcicgQU5EIHB1bGxfcmVxdWVzdF9udW1iZXIgPSAwKSBPUiBwdWxsX3JlcXVlc3RfbnVtYmVyICE9IDApCkdST1VQIEJZIGRheQpPUkRFUiBCWSBkYXkgREVTQwo=

```
File: test_external_http_authenticator/test.py:53 - in started_cluster
    run_echo_server()
File: test_external_http_authenticator/test.py:46 - in run_echo_server
    raise Exception("Echo server is not responding")
E   Exception: Echo server is not responding
===== Captured log setup =====
2025-11-14 15:36:14.660000 [ 488 ] INFO : Running tests in /home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests/integration/test_external_http_authenticator/test.py (cluster.py:3214, start)
2025-11-14 15:36:15.095000 [ 488 ] INFO : Trying to create Minio instance by command docker compose --project-name roottestexternalhttpauthenticator-gw3 --env-file /home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests/integration/test_external_http_authenticator/_instances-gw3/.env --file /home/ubuntu/actions-runner/_work/ClickHouse/ClickHouse/tests/integration/helpers/../../../tests/integration/compose/docker_compose_minio.yml --verbose up -d (cluster.py:3608, start)
2025-11-14 15:36:15.805000 [ 488 ] INFO : Trying to connect to Minio... (cluster.py:3614, start)
2025-11-14 15:36:15.810000 [ 488 ] WARNING : Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fa90645ac20>: Failed to establish a new connection: [Errno 111] Connection refused')': / (connectionpool.py:870, urlopen)
2025-11-14 15:36:15.810000 [ 488 ] WARNING : Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fa90645b370>: Failed to establish a new connection: [Errno 111] Connection refused')': / (connectionpool.py:870, urlopen)
2025-11-14 15:36:15.810000 [ 488 ] WARNING : Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fa90645a3e0>: Failed to establish a new connection: [Errno 111] Connection refused')': / (connectionpool.py:870, urlopen)
2025-11-14 15:36:16.167000 [ 488 ] WARNING : Connection dropped: socket connection error: Connection refused (connection.py:622, _connect_attempt)
2025-11-14 15:36:16.311000 [ 488 ] INFO : Connecting to 172.16.5.3(172.16.5.3):2181, use_ssl: False (connection.py:650, _connect)
2025-11-14 15:36:16.311000 [ 488 ] WARNING : Connection dropped: socket connection error: Connection refused (connection.py:622, _connect_attempt)
2025-11-14 15:36:16.523000 [ 488 ] INFO : Connecting to 172.16.5.3(172.16.5.3):2181, use_ssl: False (connection.py:650, _connect)
2025-11-14 15:36:16.523000 [ 488 ] WARNING : Connection dropped: socket connection error: Connection refused (connection.py:622, _connect_attempt)
2025-11-14 15:36:16.846000 [ 488 ] INFO : Connecting to 172.16.5.3(172.16.5.3):2181, use_ssl: False (connection.py:650, _connect)
2025-11-14 15:36:16.846000 [ 488 ] WARNING : Connection dropped: socket connection error: Connection refused (connection.py:622, _connect_attempt)
2025-11-14 15:36:17.590000 [ 488 ] INFO : Connecting to 172.16.5.3(172.16.5.3):2181, use_ssl: False (connection.py:650, _connect)
2025-11-14 15:36:17.590000 [ 488 ] WARNING : Connection dropped: socket connection error: Connection refused (connection.py:622, _connect_attempt)
2025-11-14 15:36:18.686000 [ 488 ] INFO : Connecting to 172.16.5.3(172.16.5.3):2181, use_ssl: False (connection.py:650, _connect)
2025-11-14 15:36:18.686000 [ 488 ] WARNING : Connection dropped: socket connection error: Connection refused (connection.py:622, _connect_attempt)
2025-11-14 15:36:20.804000 [ 488 ] INFO : Connecting to 172.16.5.3(172.16.5.3):2181, use_ssl: False (connection.py:650, _connect)
2025-11-14 15:36:20.805000 [ 488 ] WARNING : Connection dropped: socket connection error: Connection refused (connection.py:622, _connect_attempt)
```

--- a/tests/integration/test_external_http_authenticator/http_auth_server.py
+++ b/tests/integration/test_external_http_authenticator/http_auth_server.py
@@ -1,6 +1,7 @@
 import base64
 import http.server
 import json
+from typing import List
 
 GOOD_PASSWORD = "good_password"
 USER_RESPONSES = {
@@ -12,7 +13,8 @@ USER_RESPONSES = {
 
 
 class RequestHandler(http.server.BaseHTTPRequestHandler):
-    def decode_basic(self, data):
+    @classmethod
+    def decode_basic(cls, data: bytes) -> List[str]:
         decoded_data = base64.b64decode(data).decode("utf-8")
         return decoded_data.split(":", 1)
 
@@ -24,8 +26,6 @@ class RequestHandler(http.server.BaseHTTPRequestHandler):
 
     def do_ACCESS_GRANTED(self, user: str) -> None:
         self.send_response(http.HTTPStatus.OK)
-        body = ""
-
         response = USER_RESPONSES.get(user)
 
         if isinstance(response, dict):
@@ -35,7 +35,7 @@ class RequestHandler(http.server.BaseHTTPRequestHandler):
 
         body_raw = body.encode("utf-8")
         self.send_header("Content-Type", "application/json")
-        self.send_header("Content-Length", len(body_raw))
+        self.send_header("Content-Length", str(len(body_raw)))
         self.end_headers()
         self.wfile.write(body_raw)
 
@@ -63,7 +63,9 @@ class RequestHandler(http.server.BaseHTTPRequestHandler):
             user_name, password = self.decode_basic(data)
             if password == GOOD_PASSWORD:
                 self.do_ACCESS_GRANTED(user_name)
-            elif self.headers.get("Custom-Header") == "ok" and not self.headers.get("User-Agent"):
+            elif self.headers.get("Custom-Header") == "ok" and not self.headers.get(
+                "User-Agent"
+            ):
                 self.do_ACCESS_GRANTED(user_name)
             else:
                 self.do_AUTHHEAD()

--- a/tests/integration/test_external_http_authenticator/test.py
+++ b/tests/integration/test_external_http_authenticator/test.py
@@ -1,11 +1,10 @@
 import json
-import logging
 import os
-import sys
 
 import pytest
 
 from helpers.cluster import ClickHouseCluster
+from helpers.test_tools import wait_condition
 
 from .http_auth_server import GOOD_PASSWORD, USER_RESPONSES
 
@@ -19,12 +18,12 @@ SCRIPT_DIR = os.path.dirname(os.path.realpath(__file__))
 
 
 def run_echo_server():
-    tmp = instance.copy_file_to_container(
+    instance.copy_file_to_container(
         os.path.join(SCRIPT_DIR, "http_auth_server.py"),
         "/http_auth_server.py",
     )
 
-    tmp = instance.exec_in_container(
+    instance.exec_in_container(
         [
             "bash",
             "-c",
@@ -34,16 +33,18 @@ def run_echo_server():
         user="root",
     )
 
-    for _ in range(0, 10):
-        ping_response = instance.exec_in_container(
+    def check_server():
+        return instance.exec_in_container(
             ["curl", "-s", f"http://localhost:8000/health"],
             nothrow=True,
         )
-        logging.debug(f"Reply1: {ping_response}")
-        if ping_response == "OK":
-            return
 
-    raise Exception("Echo server is not responding")
+    wait_condition(
+        check_server,
+        lambda response: response == "OK",
+        max_attempts=20,
+        delay=0.5,
+    )
 
 
 @pytest.fixture(scope="module")
@@ -93,13 +94,25 @@ def test_basic_auth_failed(started_cluster):
         "SELECT currentUser()", user="good_user", password="bad_password"
     )
 
+
 def test_header_failed(started_cluster):
     for header_name in ["Custom-Header", "CUSTOM-HEADER", "custom-header"]:
         ping_response = instance.exec_in_container(
-            ["curl", "-s", "-u", "good_user:bad_password", "-H", f"{header_name}: ok",  "--data", f"SELECT 2+2", f"http://localhost:8123"],
+            [
+                "curl",
+                "-s",
+                "-u",
+                "good_user:bad_password",
+                "-H",
+                f"{header_name}: ok",
+                "--data",
+                f"SELECT 2+2",
+                f"http://localhost:8123",
+            ],
             nothrow=True,
         )
         assert ping_response == "4\n"
+
 
 def test_session_settings_from_auth_response(started_cluster):
     for user, response in USER_RESPONSES.items():

--- a/tests/integration/test_external_http_authenticator/test.py
+++ b/tests/integration/test_external_http_authenticator/test.py
@@ -1,5 +1,6 @@
 import json
 import os
+import typing
 
 import pytest
 
@@ -33,7 +34,7 @@ def run_echo_server():
         user="root",
     )
 
-    def check_server():
+    def check_server() -> str:
         return instance.exec_in_container(
             ["curl", "-s", f"http://localhost:8000/health"],
             nothrow=True,
@@ -48,7 +49,7 @@ def run_echo_server():
 
 
 @pytest.fixture(scope="module")
-def started_cluster():
+def started_cluster() -> typing.Generator[ClickHouseCluster, None]:
     try:
         cluster.start()
         run_echo_server()
@@ -57,7 +58,7 @@ def started_cluster():
         cluster.shutdown()
 
 
-def test_user_from_config_basic_auth_pass(started_cluster):
+def test_user_from_config_basic_auth_pass(started_cluster: ClickHouseCluster):
     assert (
         instance.query("SHOW CREATE USER good_user")
         == "CREATE USER good_user IDENTIFIED WITH http SERVER \\'basic_server\\' SCHEME \\'BASIC\\' SETTINGS PROFILE `default`\n"
@@ -70,7 +71,7 @@ def test_user_from_config_basic_auth_pass(started_cluster):
     )
 
 
-def test_user_create_basic_auth_pass(started_cluster):
+def test_user_create_basic_auth_pass(started_cluster: ClickHouseCluster):
     instance.query(
         "CREATE USER basic_user IDENTIFIED WITH HTTP SERVER 'basic_server' SCHEME 'BASIC'"
     )
@@ -89,13 +90,13 @@ def test_user_create_basic_auth_pass(started_cluster):
     instance.query("DROP USER basic_user")
 
 
-def test_basic_auth_failed(started_cluster):
+def test_basic_auth_failed(started_cluster: ClickHouseCluster):
     assert "good_user: Authentication failed" in instance.query_and_get_error(
         "SELECT currentUser()", user="good_user", password="bad_password"
     )
 
 
-def test_header_failed(started_cluster):
+def test_header_failed(started_cluster: ClickHouseCluster):
     for header_name in ["Custom-Header", "CUSTOM-HEADER", "custom-header"]:
         ping_response = instance.exec_in_container(
             [
@@ -114,7 +115,7 @@ def test_header_failed(started_cluster):
         assert ping_response == "4\n"
 
 
-def test_session_settings_from_auth_response(started_cluster):
+def test_session_settings_from_auth_response(started_cluster: ClickHouseCluster):
     for user, response in USER_RESPONSES.items():
         query_id = f"test_query_{user}"
         assert (

--- a/tests/integration/test_external_http_authenticator/test.py
+++ b/tests/integration/test_external_http_authenticator/test.py
@@ -49,7 +49,7 @@ def run_echo_server():
 
 
 @pytest.fixture(scope="module")
-def started_cluster() -> typing.Generator[ClickHouseCluster, None]:
+def started_cluster() -> typing.Generator[ClickHouseCluster, None, None]:
     try:
         cluster.start()
         run_echo_server()

